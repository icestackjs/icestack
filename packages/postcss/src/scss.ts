import { parse as scssParse, stringify as scssStringify } from 'postcss-scss'
import type { Root } from 'postcss'
import { compressCssSelector } from '@icestack/postcss-utils'

export function parse(css: string) {
  const root = scssParse(css)
  root.walkRules((rule) => {
    rule.selector = compressCssSelector(rule.selector)
  })
  return root
}

export function stringify(root: Root) {
  let result = ''
  scssStringify(root, (i) => {
    result += i
  })
  return result
}
