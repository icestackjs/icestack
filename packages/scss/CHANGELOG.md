# @icestack/scss

## 1.0.1

### Patch Changes

- [`7475342`](https://github.com/sonofmagic/icestack/commit/747534263ba1dba873afc8bd409160e189a75e54) Thanks [@sonofmagic](https://github.com/sonofmagic)! - deps on @icestack config and add more colors types

- Updated dependencies [[`7475342`](https://github.com/sonofmagic/icestack/commit/747534263ba1dba873afc8bd409160e189a75e54)]:
  - @icestack/shared@1.0.1

## 1.0.0

### Major Changes

- [#25](https://github.com/sonofmagic/icestack/pull/25) [`b7f2ee4`](https://github.com/sonofmagic/icestack/commit/b7f2ee41c516b63f6093f5c92fc2375817438347) Thanks [@sonofmagic](https://github.com/sonofmagic)! - chore: 1.0.0

### Patch Changes

- Updated dependencies [[`b7f2ee4`](https://github.com/sonofmagic/icestack/commit/b7f2ee41c516b63f6093f5c92fc2375817438347)]:
  - @icestack/shared@1.0.0
