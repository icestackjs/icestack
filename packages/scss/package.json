{
  "name": "@icestack/scss",
  "description": "The scss utils of @icestack/ui",
  "version": "1.3.8",
  "scripts": {
    "build": "unbuild",
    "test": "vitest run",
    "test:dev": "vitest",
    "release": "pnpm publish",
    "sync": "cnpm sync @icestack/scss"
  },
  "keywords": [
    "scss",
    "component",
    "generate",
    "generator",
    "tailwindcss",
    "unocss",
    "icestack",
    "ui"
  ],
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.mjs",
      "require": "./dist/index.cjs"
    }
  },
  "main": "./dist/index.cjs",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "files": [
    "dist"
  ],
  "dependencies": {
    "@icestack/shared": "workspace:^",
    "immutable": "^4.3.6",
    "sass": "^1.77.5"
  },
  "author": "SonOfMagic <qq1324318532@gmail.com>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/sonofmagic/icestack"
  },
  "homepage": "https://ui.icebreaker.top",
  "repository": {
    "type": "git",
    "url": "https://github.com/sonofmagic/icestack"
  },
  "devDependencies": {
    "@icestack/postcss-utils": "workspace:^"
  }
}
