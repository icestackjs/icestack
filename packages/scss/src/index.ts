import { OrderedMap } from 'immutable'
import * as sass from 'sass'
import { defu } from '@icestack/shared'

export interface TransformJsToSassOptions {
  quotes?: boolean
  denominatorUnits?: string[]
  numeratorUnits?: string[]
  radix?: number
}
// : sass.Value | sass.Value[] | sass.SassList | sass.SassMap
export function transformJsToSass(raw: any, options?: TransformJsToSassOptions): sass.Value {
  const { denominatorUnits, numeratorUnits, quotes, radix } = defu<Required<TransformJsToSassOptions>, TransformJsToSassOptions[]>(options, {
    quotes: false
  })
  if (raw === undefined || raw === null) {
    return sass.sassNull
  } else if (typeof raw === 'string') {
    // default quotes false
    return new sass.SassString(raw, {
      quotes
    })
  } else if (typeof raw === 'boolean') {
    return raw ? sass.sassTrue : sass.sassFalse
  } else if (typeof raw === 'number') {
    return new sass.SassNumber(raw, {
      denominatorUnits,
      numeratorUnits
    })
  } else if (typeof raw === 'bigint') {
    return new sass.SassString(raw.toString(radix), {
      quotes
    })
  } else if (Array.isArray(raw)) {
    return new sass.SassList(
      raw.map((x) => {
        return transformJsToSass(x)
      })
    )
  } else if (typeof raw === 'object') {
    return new sass.SassMap(
      OrderedMap(
        Object.entries(raw).map(([key, value]) => {
          return [transformJsToSass(key), transformJsToSass(value)]
        })
      )
    )
  } else if (typeof raw === 'function') {
    return typeof raw.signature === 'string' ? new sass.SassFunction(raw.signature, raw) : sass.sassNull
  } else {
    return sass.sassNull
  }
}

export function compileScssString(str: string, sassOptions?: sass.StringOptions<'sync'>) {
  return sass.compileString(str, sassOptions)
}
