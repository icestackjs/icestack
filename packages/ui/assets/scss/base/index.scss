@use "../utils/index.scss" as Util;
@use "../colors.scss" as Colors;

:root {
  @each $var, $color in Colors.$root-vars {
    #{$var}: Util.getRgbaString($color);
  }
}

[data-theme="dark"] {
  @each $var, $color in Colors.$dark-root-vars {
    #{$var}: Util.getRgbaString($color);
  }
}
