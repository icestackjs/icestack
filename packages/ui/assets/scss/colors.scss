$root-vars: (
  --primary-content: #ffffff,
  --primary: #1989fa,
  --success-content: #ffffff,
  --success: #07c160,
  --error-content: #ffffff,
  --error: #ee0a24,
  --warning-content: #ffffff,
  --warning: #ff976a
);

