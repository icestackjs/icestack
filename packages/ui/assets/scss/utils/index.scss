@use "sass:string";
@use "sass:math";
@use "sass:meta";
@use "sass:map";
// https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/rgb()
@function getRgbaString($color) {
  @if (meta.type-of($color) == color) {
    $alpha: color.alpha($color);

    @if ($alpha == 1 or $alpha == 0) {
      @return #{color.red($color) color.green($color) color.blue($color)};
    } @else {
      @return #{color.red($color) color.green($color) color.blue($color) "/" color.alpha($color)};
    }
  } @else {
    @return $color;
  }
}

@function getRgbString($color) {
  @if (meta.type-of($color) == color) {
    @return #{color.red($color) color.green($color) color.blue($color)};
  } @else {
    @return $color;
  }
}

$globalAtMediaHover: globalAtMediaHover();

@function getSelector($type, $prefix: "-") {
  @if ($type == "") {
    @return "";
  } @else {
    @return "#{$prefix}#{$type}";
  }
}

@function removeBorderPrefix($str) {
  @return removePrefix($str, "border-");
}

@function removeBgPrefix($str) {
  @return removePrefix($str, "bg-");
}

@function removeTextPrefix($str) {
  @return removePrefix($str, "text-");
}

@function removePrefix($str, $prefix) {
  @return string.slice($str, string.length($prefix) + 1);
}

@mixin AtMediaHover {
  @if ($globalAtMediaHover) {
    @media (hover: hover) {
      @content;
    }
  } @else {
    @content;
  }
}

@function ApplyAtMediaHover($str) {
  @if ($globalAtMediaHover) {
    @return "[@media(hover:hover)]:#{$str}";
  } @else {
    @return $str;
  }
}

@function composeStringArray($list) {
  $result: null;
  @each $item in $list {
    @if ($result == null) {
      $result: $item;
    } @else {
      $result: #{$result} #{$item};
    }
  }

  @return #{$result};
}

// type-of null == "null"
// null == null

// @else if($type == list) {
//   $arg: (
//     apply: composeStringArray($arg)
//   );
// }
@mixin resolve($arg) {
  $type: meta.type-of($arg);

  @if ($type == string) {
    $arg: (
      apply: $arg
    );
  } @else if($type == map and meta.type-of(map.get($arg, "css")) == null) {
    $arg: (
      css: $arg
    );
  } @else if($type == "null") {
    $arg: ();
  }
  $apply: map.get($arg, "apply");
  $css: map.get($arg, "css");
  $sort: map.get($arg, "sort");
  @if ($sort == null or $sort == "asc") {
    @if ($apply != null and meta.type-of($apply) == string) {
      @apply #{$apply};
    }
    @if ($css!=null and meta.type-of($css) == map) {
      @each $name, $value in $css {
        #{$name}: $value;
      }
    }
  } @else if($sort == "desc") {
    @if ($css!=null and meta.type-of($css) == map) {
      @each $name, $value in $css {
        #{$name}: $value;
      }
    }
    @if ($apply != null and meta.type-of($apply) == string) {
      @apply #{$apply};
    }
  }
}

@mixin spread($map, $depth: 0) {
  @if (meta.type-of($map) == map) {
    @each $s, $m in $map {
      @if ($s == "css" or $s == "apply" or $s == "sort") {
        @if ($depth > 0) {
          @if ($s == "apply") {
            @if ($m != null and meta.type-of($m) == string) {
              @apply #{$m};
            }
          } @else if($s == "css") {
            @if ($m!=null and meta.type-of($m) == map) {
              @each $name, $value in $m {
                #{$name}: $value;
              }
            }
          }
        }
      } @else {
        #{$s} {
          @include spread($m, $depth + 1);
        }
      }
    }
  }
}
