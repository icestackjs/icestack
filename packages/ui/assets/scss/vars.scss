// 统一加变量前缀

$root-vars: (
  --primary: #1677ff,
  // --primary-selected: #e6f4ff,
  --primary-focus: #4096ff,
  // --primary-click: #0958d9,
  --primary-content: #ffffff,
  // ---
  --success-content: #ffffff,
  --success: #07c160,
  --success-focus: #73d13d,

  --error-content: #ffffff,
  --error: #ee0a24,
  --error-focus: #ff4d4f,

  --warning-content: #ffffff,
  --warning: #ff976a,
  --warning-focus: #ffc53d,

  --info: #1677ff,
  --info-content: #ffffff,
  --info-focus: #4096ff,

  --neutral: rgb(43, 52, 64),
  --neutral-content: rgb(216, 221, 228),
  --neutral-focus: rgb(36, 43, 51),
  // --- Neutral Color Start ---
  // 900
  --heading-text: #171717,
  // 700
  --text-100: #171717,
  --text-200: #404040,

  --base-100: #ffffff,
  --base-200: rgb(242, 242, 242),
  --base-300: rgb(229, 230, 230),
  --base-content: rgb(31, 41, 55),
  --disabled-text: #00000040,
  --default-border: #d9d9d9,
  --separator: #0505050f,
  --layout-background: #f5f5f5,
  // ---
  --rounded-box: 1rem,
  --rounded-btn: 0.5rem,
  --rounded-badge: 1.9rem,
  --animation-btn: 0.25s,
  --animation-input: 0.2s,
  --btn-text-case: uppercase,
  --btn-focus-scale: 0.95,
  --border-btn: 1px,
  --tab-border: 1px,
  --tab-radius: 0.5rem
);

// $dark-root-vars: (
//   --primary: #1668dc,
//   // --primary-selected: #111a2c,
//   --primary-focus: #1554ad,
//   // --primary-click: #3c89e8,
//   --primary-content: #ffffff,
//   // ---
//   --success-content: #ffffff,
//   --success: #49aa19,
//   --error-content: #ffffff,
//   --error: #d32029,
//   --warning-content: #ffffff,
//   --warning: #d89614,

//   --info: #1668dc,
//   --info-content: #ffffff,

//   --neutral: rgb(42, 50, 60),
//   --neutral-content: rgb(166, 173, 186),
//   // --- Neutral Color Start ---
//   --heading-text: #e5e5e5,
//   --text-100: #e5e5e5,
//   // 400
//   --text-200: #a3a3a3,

//   --base-100: rgb(29, 35, 42),
//   --base-200: rgb(25, 30, 36),
//   --base-300: rgb(21, 25, 30),
//   --base-content: rgb(166, 173, 186),
//   --disabled-text: #ffffff40,
//   --default-border: #424242,
//   --separator: #fdfdfd1f,
//   --layout-background: #000000,

//   --rounded-box: 1rem,
//   --rounded-btn: 0.5rem,
//   --rounded-badge: 1.9rem,
//   --animation-btn: 0.25s,
//   --animation-input: 0.2s,
//   --btn-text-case: uppercase,
//   --btn-focus-scale: 0.95,
//   --border-btn: 1px,
//   --tab-border: 1px,
//   --tab-radius: 0.5rem
// );
