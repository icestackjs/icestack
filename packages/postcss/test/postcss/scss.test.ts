import type { Node, Rule, ChildNode, Root, AtRule, Declaration, Comment } from 'postcss'
import { parse, stringify } from '@/scss'
import { compressCssSelector } from '@/index'

function getKey(node: AtRule | Rule | Root | Comment | Declaration) {
  switch (node.type) {
    case 'root': {
      return '#'
    }
    case 'atrule': {
      return '@' + node.name + ' ' + node.params
    }
    case 'rule': {
      return compressCssSelector(node.selector)
    }
    // No default
  }
  return '$'
}

function calcPath(node: AtRule | Rule): string[] {
  const paths: string[] = [getKey(node)]

  let parent: AtRule | Rule | Root = node.parent as AtRule | Rule | Root
  while (parent) {
    paths.push(getKey(parent as AtRule | Rule | Root))
    parent = parent.parent as AtRule | Rule | Root
  }
  return paths
}

// 从右到左
function merge(leftNode: AtRule | Rule | Root, rightNode: AtRule | Rule | Root) {
  // leftNode
  for (const right of rightNode.nodes) {
    let insertFlag = true
    for (const left of leftNode.nodes) {
      if (left.type === 'rule' && right.type === 'rule' && compressCssSelector(left.selector) === compressCssSelector(right.selector)) {
        merge(left, right)
        insertFlag = false
      } else if (left.type === 'atrule' && right.type === 'atrule' && left.params === right.params && left.name === right.name) {
        if (left.name === 'apply' && right.name === 'apply') {
          left.after(right)
        } else {
          merge(left, right)
        }
        insertFlag = false
      }
      // else if(left.type === 'decl' && right.type === 'decl'){
      //   if(left.prop === )
      // }
    }
    insertFlag && leftNode.push(right)
  }
}

describe('scss', () => {
  it('parse case 0', () => {
    const root = parse(` .avatar{
      @apply relative inline-flex;
    }`)

    const mergeItem = parse(` .avatar{
      @apply bg-xx;
    }`)
    merge(root, mergeItem)

    expect(root.toString()).toMatchSnapshot()
  })

  it('parse case 1', () => {
    const root = parse(` .avatar{
      @apply relative inline-flex;
    }`)

    const mergeItem = parse(` .avatar1{
      @apply bg-xx;
    }`)
    merge(root, mergeItem)

    expect(root.toString()).toMatchSnapshot()
  })

  it('parse case 2', () => {
    const root = parse(` .avatar{
      @apply relative inline-flex;
      color: blue;
    }`)

    const mergeItem = parse(` .avatar{
      @apply bg-xx;
      color: red;
    }`)
    merge(root, mergeItem)

    expect(root.toString()).toMatchSnapshot()
  })
  it('parse case 10', () => {
    const root = parse(` .avatar{
      @apply relative inline-flex;
      > div {
        @apply block aspect-square overflow-hidden;
      }
      img {
        @apply h-full w-full object-cover;
      }
      &.placeholder{
        > div{
          @apply flex items-center justify-center;
        }
      }
    }`)

    const mergeItem = parse(` .avatar{
      @apply bg-xx;
      > div {
        @apply mt-2;
        color:red;
      }
      &.placeholder{
        color:blue;
        > span{
          @apply flex items-center justify-center;
        }
        >     div{
          @apply text-xs;
        }
        >   div{
          @apply text-lg;

          bg: red;
        }
      }
    }`)
    merge(root, mergeItem)
    // const mkm = new Map()

    // root.walkAtRules((atRule) => {
    //   if (atRule.name === 'apply') {
    //     mkm.set(calcPath(atRule).join('|'), atRule)
    //   }
    // })

    // root.walkRules((rule) => {
    //   mkm.set(calcPath(rule).join('|'), rule)
    // })
    // override
    // root.each(mergeItem)
    // mergeItem.walkAtRules((atRule) => {
    //   if (atRule.name === 'apply') {
    //     const n = mkm.get(calcPath(atRule).join('|'))
    //     // mkm.set(, atRule)
    //   }
    // })

    // mergeItem.walkRules((rule) => {
    //   // mkm.set(calcPath(rule).join('|'), rule)
    // })

    expect(root.toString()).toMatchSnapshot()
  })
})
