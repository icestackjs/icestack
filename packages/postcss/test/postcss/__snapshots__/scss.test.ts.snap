// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`scss > parse case 0 1`] = `
" .avatar{
      @apply relative inline-flex;
      @apply bg-xx;
    }"
`;

exports[`scss > parse case 1 1`] = `
" .avatar{
      @apply relative inline-flex;
    } .avatar1{
      @apply bg-xx;
    }"
`;

exports[`scss > parse case 2 1`] = `
" .avatar{
      @apply relative inline-flex;
      color: blue;
      @apply bg-xx;
      color: red;
    }"
`;

exports[`scss > parse case 10 1`] = `
" .avatar{
      @apply relative inline-flex;
      >div {
        @apply block aspect-square overflow-hidden;
        @apply mt-2;
        color:red;
      }
      img {
        @apply h-full w-full object-cover;
      }
      &.placeholder{
        >div{
          @apply flex items-center justify-center;
          @apply text-xs;
          @apply text-lg;

          bg: red;
        }
        color:blue;
        >span{
          @apply flex items-center justify-center;
        }
      }
      @apply bg-xx
    }"
`;
